mutexLock/mutexUnlock.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package mutexLock
+
+import (
+	"go/ast"
+	"go/token"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "mutexLock",
+	Doc:  "check for cases where a mutex is left locked before returning",
+	Run:  run,
+}
+
+func isMutex(varType string) bool {
+	switch varType {
+	case "sync.Mutex", "sync.RWMutex":
+		return true
+	}
+
+	return false
+}
+
+func isLock(methodName string) bool {
+	switch methodName {
+	case "Lock", "RLock":
+		return true
+	}
+
+	return false
+}
+
+func isUnlock(methodName string) bool {
+	switch methodName {
+	case "Unlock", "RUnlock":
+		return true
+	}
+
+	return false
+}
+
+type checkState struct {
+	lockPos   token.Pos
+	unlockPos token.Pos
+	counter   int
+	reported  bool
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	for _, file := range pass.Files {
+		var m map[string]*checkState
+
+		check := func() {
+			for k, v := range m {
+				if v.counter > 0 {
+					if !v.reported {
+						pass.Reportf(v.lockPos, "possible return with mutex %s locked", k)
+						v.reported = true
+					}
+				}
+			}
+		}
+
+		ast.Inspect(file, func(node ast.Node) bool {
```

Comment:
There's also a cool way to filter by only specific node types. Since we are interested in only FuncDecl, ReturnStmt and CallExpr, we can do:

```go
inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
	// We filter only function calls.
	nodeFilter := []ast.Node{
		(*ast.FuncDecl)(nil),
		(*ast.ReturnStmt)(nil),
		(*ast.CallExpr)(nil),
	}

	inspect.Preorder(nodeFilter, func(n ast.Node) {
		switch n.(type) {} // handle only the 3 cases.
	})
```

Reply:
Nice! To be honest I had no idea which nodes I needed when I started but this could be a nice refactor.

Reply:
TIL :)